File: python/909. Snakes and Ladders.py
class Solution(object):
    def snakesAndLadders(self, board):
        """
        :type board: List[List[int]]
        :rtype: int
        """
        def coordinate(pos, N):
            pos -= 1
            x = pos / N
            if x&1:
                y = N - pos%N - 1
            else:
                y = pos%N

            return N-1-x, y


        N = len(board)
        step = 0
        dis = [float('inf')]*(N*N + 1)
        dis[1] = 0
        que = [1]
        while que:
            sz = len(que)
            while sz:
                sz -= 1
                pos = que.pop(0)
                if pos >= N*N:
                    return dis[pos]
                i = 1
                while i<=6 and pos+i<=N*N:
                    x, y = coordinate(pos+i, N)
                    npos = pos+i if board[x][y]==-1 else board[x][y]
                    if dis[pos] + 1 < dis[npos]:
                        dis[npos] = dis[pos] + 1
                        que.append(npos)
                    i += 1


        return -1
                    




if __name__ == '__main__':
    # N = 6
    # pos = 33
    # print coordinate(pos, N)
    board = [[-1,-1,-1,46,47,-1,-1,-1],
            [51,-1,-1,63,-1,31,21,-1],
            [-1,-1,26,-1,-1,38,-1,-1],
            [-1,-1,11,-1,14,23,56,57],
            [11,-1,-1,-1,49,36,-1,48],
            [-1,-1,-1,33,56,-1,57,21],
            [-1,-1,-1,-1,-1,-1,2,-1],
            [-1,-1,-1,8,3,-1,6,56]]

    su = Solution()
    print su.snakesAndLadders(board)

File: go/src/leetcode/909. Snakes and Ladders/909. Snakes and Ladders.go
package main

import "fmt"

func snakesAndLadders(board [][]int) int {
	N := len(board)
	N2 := N*N
	visited := make([]bool, N2)
	que := []int{0}
	visited[0] = true
	var step int
	for len(que) > 0 {
		for sz := len(que); sz > 0; sz-- {
			i := que[0]
			que = que[1:]

			if i == N2-1 {
				return step
			}
			for k := 1; k <= 6 && i+k < N2; k++ {
				ni := i + k
				if visited[ni] {
					continue
				}
				visited[ni] = true
				var x, y int
				x = N - 1 - ni/N
				if (N - 1 - x)&1 == 1 {
					y = N - 1 - ni%N
				} else {
					y = ni%N
				}
				if board[x][y] != -1 {
					ni = board[x][y]-1
				}
				que = append(que, ni)
			}
		}
		step++
	}
	return -1
}

func main() {
	fmt.Println(snakesAndLadders([][]int{
		{-1,-1,-1,-1,-1,-1},
		{-1,-1,-1,-1,-1,-1},
		{-1,-1,-1,-1,-1,-1},
		{-1,35,-1,-1,13,-1},
		{-1,-1,-1,-1,-1,-1},
		{-1,15,-1,-1,-1,-1},
	}))

	fmt.Println(snakesAndLadders([][]int{
		{-1,-1,-1},
		{-1,9,8},
		{-1,8,9},
	}))

	fmt.Println(snakesAndLadders([][]int{
		{-1,-1,19,10,-1},
		{2,-1,-1,6,-1},
		{-1,17,-1,19,-1},
		{25,-1,20,-1,-1},
		{-1,-1,-1,-1,15},
	}))

}